hub/core/meta/encode/shape.py

```diff
@@ -1,145 +1,95 @@
-
-    @property
-    def nbytes(self):
-        return self._encoded_shapes.nbytes
-
-    @property
-    def array(self):
-        return self._encoded_shapes
-
-    @property
-    def num_samples(self) -> int:
-        if len(self._encoded_shapes) == 0:
-            return 0
-        return int(self._encoded_shapes[-1, -1] + 1)
-
-    def add_shape(
-        self,
-        shape: Tuple[int],
-        count: int,
-    ):
-
-        if count <= 0:
-            raise ValueError(f"Shape `count` should be > 0. Got {count}.")
-
-        if self.num_samples != 0:
-            last_shape = self[-1]
+class ShapeEncoder(Encoder):
+    """Custom compressor that allows reading of shapes from a sample index without decompressing.
+    Requires that all shapes encoded have the same dimensionality (`len(shape)`).
+
+    Layout:
+        `_encoded` is a 2D array.
+
+        Rows:
+            The number of rows is equal to the number of unique runs of shapes that exist upon ingestion. See examples below.
+
+        Columns:
+            The number of columns is equal to the dimensionality (`len(shape)`) of the shapes + 1.
+            Each row looks like this: [shape_dim0, shape_dim1, shape_dim2, ..., last_index], where `last_index`
+            is equal to the last index the specified shape in that row exists. This means that a shape can be shared
+            by multiple samples, so long as they were added directly after each other. See examples below.
+
+        Fixed Example:
+            >>> enc = ShapeEncoder()
+            >>> enc.register_samples((1,), 100)  # represents scalar values
+            >>> enc._encoded
+            [[1, 99]]
+            >>> enc.register_samples((1,), 10000)
+            >>> enc._encoded
+            [[1, 10099]]
+            >>> enc.num_samples
+            10100
+            >>> enc[5000]
+            (1,)
+
+        Dynamic Example:
+            >>> enc = ShapeEncoder()
+            >>> enc.register_samples((28, 28), 1)
+            >>> enc._encoded
+            [[28, 28, 0]]
+            >>> enc.register_samples((28, 28, 10))
+            >>> enc._encoded
+            [[28, 28, 10]]
+            >>> enc.register_samples((29, 28, 5))
+            >>> enc._encoded
+            [[28, 28, 10],
+             [29, 28, 15]]
+            >>> enc.register_samples((28, 28, 3))
+            >>> enc._encoded
+            [[28, 28, 10],
+             [29, 28, 15],
+             [28, 28, 18]]
+            >>> enc.num_samples
+            19
+            >>> enc[10]
+            (28, 28)
+            >>> enc[11]
+            (29, 28)
+
+        Best case scenario:
+            The best case scenario is when all samples have the same shape. This means that only 1 row is created.
+            This is O(1) lookup.
+
+        Worst case scenario:
+            The worst case scenario is when all samples have different shapes. This means that there are as many rows as there are samples.
+            This is O(log(N)) lookup.
```

Comment:
How much of this can be factored elsewhere? I've seen this same snippet several times by now. Why is this not in the encoder base class?

Reply:
generalized docs